File: SistemaGestorDeLibros.java
import java.io.*;
import java.util.Scanner;

public class SistemaGestorDeLibros {

    private static void capturaDatos(listaDinamica_Libro coleccion, String ubicacion) {
        Scanner lectura = new Scanner(System.in);
        Libro libroNuevo = new Libro();
        // ------- ------- ------- Captura
        System.out.println("_______Ingrese los datos del libro_______");
        System.out.print("Titulo del libro: ");
        libroNuevo.setNombre(lectura.nextLine());
        System.out.print("Autor: ");
        libroNuevo.setAutor(lectura.nextLine());
        System.out.print("Precio de compra: ");
        libroNuevo.setPrecio(lectura.nextDouble());
        System.out.print("Fecha de publicacion: ");
        libroNuevo.setPublicacion(lectura.nextInt());
        // ------- ------- ------- Incersión
        switch (ubicacion) {
            case "i":
                coleccion.agregaInicio(libroNuevo);
                break;
            case "f":
                coleccion.agregaFinal(libroNuevo);
                break;
            default:
                // Si llegó a este punto, es un ERROR.
                break;
        }
    }

    public static void main(String[] args) {
        Scanner scan = new Scanner(System.in);
        int contador= -1;
        // -------------------------------------------------------------------------
        try ( FileWriter fw = new FileWriter("biblioteca.txt", true) ){            
            fw.close();
            contador= 0;
        }catch(Exception e){
            e.printStackTrace();
        }
        // -------------------------------------------------------------------------
        try ( BufferedReader bfr = new BufferedReader(new FileReader("biblioteca.txt")) ){
            String cadena = bfr.readLine();
            contador++;
            while( cadena != null ){
                cadena= bfr.readLine(); // Se cuentan las lineas en general y luego se divide por el número de campos
                contador++;
            }
            contador /= 4;
        }catch(Exception e){
            e.printStackTrace();
        }
        // -------------------------------------------------------------------------
        if( contador > 0){
            System.out.println("Registros de libros encontrados: " + contador);
        } else {
            System.out.println("Nose encontraron registros de libros.");
        }
        System.out.println("------- Oprima ENTER para continuar -------");
        scan.nextLine();
        // -------------------------------------------------------------------------
        listaDinamica_Libro coleccion = new listaDinamica_Libro();
        int opcion;
        int exit = 18;
        do {
            // Menu
            System.out.print("\033[H\033[2J");
            System.out.println("------- ------- ------- ------- -------");
            System.out.println("Sistema Gestor de Libros.");
            System.out.println(" 1 | Agregar  | Inicio.");
            System.out.println(" 2 | Agregar  | Final.");
            System.out.println(" 3 | Eliminar | Primer libro.");
            System.out.println(" 4 | Eliminar | Ultimo libro.");
            System.out.println(" 5 | Imprimir | Cantidad de libros.");
            System.out.println(" 6 | Imprimir | Lista completa.");
            System.out.println(" 7 | Imprimir | Datos de un libro.");
            System.out.println(" 8 | Ordenar  | Titulo.");
            System.out.println(" 9 | Ordenar  | Autor.");
            System.out.println("10 | Ordenar  | Precio Menor a Mayor.");
            System.out.println("11 | Ordenar  | Precio Mayor a Menor.");
            System.out.println("12 | Buscar   | Libro por Titulo.");
            System.out.println("13 | Buscar   | Libro por Autor.");
            System.out.println("14 | Buscar   | Libros por Titulo.");
            System.out.println("15 | Buscar   | Libros por Autor.");
            System.out.println("16 | Archivo  | Cargar registros.");
            System.out.println("17 | Archivo  | Guardar registros");
            System.out.println("18 | Salir del programa.");
            System.out.println("------- ------- ------- ------- -------");
            System.out.print("Accion a realizar: ");
            opcion = scan.nextInt();
            scan.nextLine();
            // -----------------
            switch (opcion) {
                case 1:
                    capturaDatos(coleccion, "i");
                    System.out.println("Se agregó el libro!");
                    break;
                case 2:
                    capturaDatos(coleccion, "f");
                    System.out.println("Se agregó el libro!");
                    break;
                case 3:
                    if (!coleccion.isEmpty()) {
                        coleccion.eliminaInicio();
                        System.out.println("Se eliminó el libro!");
                    } else {
                        System.out.println("La lista esta vacia.");
                    }
                    break;
                case 4:
                    if (!coleccion.isEmpty()) {
                        coleccion.eliminaFinal();
                        System.out.println("Se eliminó el libro!");
                    } else {
                        System.out.println("La lista esta vacia.");
                    }
                    break;
                case 5:
                    System.out.println("Elementos: " + coleccion.getElementos() + ".\n");
                    break;
                case 6:
                    if (!coleccion.isEmpty()) {
                        for (int i = 0; i < coleccion.getElementos(); i++) {
                            System.out.println(coleccion.getDatosLibro(i));
                        }
                        System.out.println("------- Oprima ENTER para continuar -------");
                        scan.nextLine();
                    } else {
                        System.out.println("No hay libros en la coleccion");
                    }
                    break;
                case 7:
                    int indice;
                    System.out.println("Ingrese el indice del libro: ");
                    indice = scan.nextInt();
                    if (indice > -1 && indice < coleccion.getElementos()) {
                        Libro accedido;
                        accedido = coleccion.getLibro(indice);
                        System.out.println("Título: " + accedido.getNombre());
                        System.out.println("Autor: " + accedido.getAutor());
                        System.out.println("Año: " + accedido.getPublicacion());
                        System.out.println("Precio: $ " + accedido.getPrecio());
                    }
                    break;
                case 8: {
                    coleccion.ordenarTitulo();
                    System.out.println("Se ordenó la lista dinámica!");
                    break;
                }
                case 9: {
                    coleccion.ordenarAutor();
                    System.out.println("Se ordenó la lista dinámica!");
                    break;
                }
                case 10: {
                    coleccion.ordenarPrecioAsc();
                    System.out.println("Se ordenó la lista dinámica!");
                    break;
                }
                case 11: {
                    coleccion.ordenarPrecioDes();
                    System.out.println("Se ordenó la lista dinámica!");
                    break;
                }
                case 12: {
                    System.out.println("Nombre de libro a buscar (coincidencia exacta): ");
                    Libro libro = coleccion.buscarLibro_Nombre(scan.nextLine());
                    if (libro.getNombre().compareTo("error") != 0) {
                        System.out.println("----------------------------------------------");
                        System.out.println("Titulo: " + libro.getNombre());
                        System.out.println("Autor: " + libro.getAutor());
                        System.out.println("Publicacion: " + libro.getPublicacion());
                        System.out.println("Precio: $" + libro.getPrecio());
                        System.out.println("----------------------------------------------");
                    } else {
                        System.out.println("No se encontro el libro.");
                    }
                    break;
                }
                case 13: {
                    System.out.println("Nombre del autor del libro (coincidencia exacta): ");
                    Libro libro = coleccion.buscarLibro_Autor(scan.nextLine());
                    if (libro.getAutor().compareTo("error") != 0) {
                        System.out.println("----------------------------------------------");
                        System.out.println("Titulo: " + libro.getNombre());
                        System.out.println("Autor: " + libro.getAutor());
                        System.out.println("Publicacion: " + libro.getPublicacion());
                        System.out.println("Precio: $" + libro.getPrecio());
                        System.out.println("----------------------------------------------");
                    } else {
                        System.out.println("No se encontro ningun libro de ese autor.");
                    }
                    break;
                }
                case 14: {
                    System.out.println("Nombre de libro a buscar (coincidencia exacta): ");
                    listaDinamica_Libro encontrados = coleccion.buscarLibros_Nombre(scan.nextLine());
                    if (encontrados.getElementos() > 0) {
                        // titulo 50 autor 30 fecha 10 precio 10
                        System.out.println("------- ------- ------- ------- -------");
                        System.out.print(String.format("%1$-50s", "Titulo"));
                        System.out.print(String.format("%1$-30s", "Autor"));
                        System.out.print(String.format("%1$-10s", "Fecha"));
                        System.out.println(String.format("%1$-10s", "Precio"));
                        for (int i = 0; i < encontrados.getElementos(); i++) {
                            Libro libro = encontrados.getLibro(i);
                            System.out.print(String.format("%1$-50s", libro.getNombre()));
                            System.out.print(String.format("%1$-30s", libro.getAutor()));
                            System.out.print(String.format("%1$-10s", libro.getPublicacion()));
                            System.out.println(String.format("%1$-10s", libro.getPrecio()));
                        }
                        System.out.println("------- Oprima ENTER para continuar -------");
                        scan.nextLine();
                    } else {
                        System.out.println("No se encontraron coincidencias con el libro.\n");
                    }
                    break;
                }
                case 15: {
                    System.out.println("Nombre del autor del libro a buscar (coincidencia exacta): ");
                    listaDinamica_Libro encontrados = coleccion.buscarLibros_Autor(scan.nextLine());
                    if (encontrados.getElementos() > 0) {
                        System.out.println("------- ------- ------- ------- -------");
                        System.out.print(String.format("%1$-50s", "Titulo"));
                        System.out.print(String.format("%1$-30s", "Autor"));
                        System.out.print(String.format("%1$-10s", "Fecha"));
                        System.out.println(String.format("%1$-10s", "Precio"));
                        for (int i = 0; i < encontrados.getElementos(); i++) {
                            Libro libro = encontrados.getLibro(i);
                            System.out.print(String.format("%1$-50s", libro.getNombre()));
                            System.out.print(String.format("%1$-30s", libro.getAutor()));
                            System.out.print(String.format("%1$-10s", libro.getPublicacion()));
                            System.out.println(String.format("%1$-10s", libro.getPrecio()));
                        }
                        System.out.println("------- Oprima ENTER para continuar -------");
                        scan.nextLine();
                    } else {
                        System.out.println("No se encontraron coincidencias con el autor.\n");
                    }
                    break;
                }
                case 16: {                    
                    try ( BufferedReader bfr = new BufferedReader(new FileReader("biblioteca.txt")) ){
                        for(int i=0; i<contador; i++){
                            String cadena1, cadena2, cadena3, cadena4;
                            cadena1= bfr.readLine();
                            cadena2= bfr.readLine();
                            cadena3= bfr.readLine();
                            cadena4= bfr.readLine();
                            Libro nuevo= new Libro();
                            if( !cadena1.isEmpty() ){
                                nuevo.setNombre(cadena1);
                                nuevo.setAutor(cadena2);
                                nuevo.setPublicacion(  Integer.parseInt(cadena3));
                                nuevo.setPrecio( Double.parseDouble(cadena4) );
                                coleccion.agregaFinal(nuevo);
                            }
                        }
                        System.out.println("Se recuperaron los registros almacenados en el archivo.");
                    }catch(Exception e){
                        e.printStackTrace();
                    }
                    break;
                }
                case 17: {
                    if(coleccion.getElementos() > 0){
                        try ( BufferedWriter archivo = new BufferedWriter(new FileWriter("biblioteca.txt")) ){
                            for(int i= 0; i<coleccion.getElementos(); i++){
                                Libro tmp_libro= coleccion.getLibro(i);
                                archivo.write( tmp_libro.getNombre() + "\n");
                                archivo.write( tmp_libro.getAutor() + "\n" );
                                archivo.write( tmp_libro.getPublicacion() + "\n" );
                                archivo.write( String.valueOf(tmp_libro.getPrecio()) + "\n" );
                            }
                            System.out.println("Se guardaron los registros en el archivo.");
                        }catch(Exception e){
                            e.printStackTrace();
                        }
                    }else{
                        System.out.print("No hay registros en la Lista Dinamica.");
                    }
                    break;
                }
                case 18:
                    System.out.println("Finalizando el programa!");
                    break;
                default:
                    System.out.println("Error en la opcion ingresada.");
            }
            // Ligero detenimiento
            try {
                Thread.sleep(5000);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        } while (opcion != exit);
        scan.close();
    }
}

File: SistemaGestorDeLibros.cpp
// SGL - SISTEMA GESTOR DE LIBROS

#include <iostream>
#include <cstdlib>
#include <iomanip> 
#include "listaDinamica_Libro.h"
using namespace std;

Libro capturaLibro(){
    cout << "Captura de datos del libro.\n\n";
    string tmp_string;
	int tmp_int;
	double tmp_double;
	Libro tmp_libro;
    cout << "Titulo: ";
	getline( cin, tmp_string);
	tmp_libro.setNombre(tmp_string);    
    cout << "Autor: ";
	getline( cin, tmp_string);
	tmp_libro.setAutor(tmp_string);
    cout << "Fecha publicacion: ";
	cin >> tmp_int;
	cin.ignore();
	tmp_libro.setPublicacion(tmp_int);
    cout << "Precio compra: ";
	cin >> tmp_double;
	cin.ignore();
	tmp_libro.setPrecio(tmp_double);
    return tmp_libro;
}

void encabezadoTabla(){
        cout << left << setw(30) << "Autor";
        cout << left << setw(80) << "Nombre del libro";
        cout << left << setw(15) << "Fecha";
        cout << left << setw(15) << "Precio";
        cout << "\n";        
        cout << left << setw(140) << setfill('-') << "-";
        cout << setfill(' ');
        cout << "\n";
}

void filaTabla(Libro libro){
        cout << setw(30) << libro.getAutor();
        cout << setw(80) << libro.getNombre();
        cout << setw(15) << libro.getPublicacion();
        cout << setw(15) << libro.getPrecio() << "\n"; 
}

void pieTabla(){
        cout << left << setw(140) << setfill('-') << "-";
        cout << setfill(' ');
        cout << "\n";
}

int main() {
    ListaDinamica_Libro misLibros;
    // ---------------------------------------------------------------------------
    int opc= 0;
    int estilo= 0;
    int exit= 17;
    do {
        system("cls");
        if(estilo){ // Estilo reducido - 1
            cout << "---------------------------------------------\n";
            cout << "Agregar  | (1) Inicio (2) Final.\n";
            cout << "Eliminar | (3) Primer libro (4) Ultimo libro.\n";
            cout << "Imprimir | (5) Cantidad de libros (6) Lista (7) Datos de un libro.\n";
            cout << "Ordenar  | (8) Titulo (9) Autor (10) Precio Menor a Mayor (11) Precio Mayor a Menor.\n";
            cout << "Menu     | (12) Cambiar a menu expandido.\n";
            cout << "Buscar   | (13) Libro x titulo (14) Libro x autor (15) Libros x titulo (16) Libros x autor\n";
            cout << "Salir    | (17) Sin guardar cambios.\n";
            cout << "---------------------------------------------\n";
            cout << "Opcion: ";
        }else{ // Estilo expandido - 0
            cout << " 1 | Agregar  | Inicio.\n";
            cout << " 2 | Agregar  | Final.\n";
            cout << " 3 | Eliminar | Primer libro.\n";
            cout << " 4 | Eliminar | Ultimo libro.\n";
            cout << " 5 | Imprimir | Cantidad de libros.\n";
            cout << " 6 | Imprimir | Lista completa.\n";
            cout << " 7 | Imprimir | Datos de un libro.\n";
            cout << " 8 | Ordenar  | Titulo.\n";
            cout << " 9 | Ordenar  | Autor.\n";
            cout << "10 | Ordenar  | Precio Menor a Mayor.\n";
            cout << "11 | Ordenar  | Precio Mayor a Menor.\n";
            cout << "12 | Menu     | Cambiar a menu reducido.\n";
            cout << "13 | Buscar   | Libro por Titulo.\n";
            cout << "14 | Buscar   | Libro por Autor.\n";
            cout << "15 | Buscar   | Libros por Titulo.\n";
            cout << "16 | Buscar   | Libros por Autor.\n";
            cout << "17 | Salir del programa.\n";
            cout << "---------------------------------------------\n";
            cout << "Opcion: ";
        }
        cin >> opc;
        cin.ignore();
        cout << "---------------------------------------------\n\n";
        switch(opc){
            case(1): {
                Libro nuevoLibro;
				nuevoLibro= capturaLibro();
                misLibros.agregaNodoInicio(nuevoLibro);
                break;
			}
            case(2): {
                Libro nuevoLibro;
				nuevoLibro= capturaLibro();
                misLibros.agregaNodoFinal(nuevoLibro);
                break;
			}
            case(3): {
                if( misLibros.elementos() ){
                    Libro retirado= misLibros.eliminaNodoInicio();
                    cout << "Se elimino el libro \"" << retirado.getNombre() << "\"\n";
                }else{
                    cout << "La lista esta vacia.\n";
                }
                break;
			}
            case(4): {
                if( misLibros.elementos() ){
                    Libro retirado= misLibros.eliminaNodoFinal();
                    cout << "Se elimino el libro \"" << retirado.getNombre() << "\"\n";
                }else{
                    cout << "La lista esta vacia.\n";
                }
                break;
			}
            case(5): {
                cout << "Libros contenidos en la lista: ";
                cout <<  misLibros.elementos() << "\n";
                break;
			}
            case(6): {
                if( misLibros.elementos() ){
                    encabezadoTabla();
                    for(int i= 0; i<misLibros.elementos(); i++){
                        Libro tmp_libro= misLibros.datoNodo(i);
                        filaTabla(tmp_libro);
					}
                    pieTabla();
				} else {
                    cout << "No hay libros en la lista.\n"; 
				}                
                break;
			}
            case(7): {
                if( misLibros.elementos() ){
                    int num;
                    cout << "Ingrese el indice del elemento: ";
                    cin >> num;
                    if(num < 0 || num >= misLibros.elementos()) break;
                    Libro libro= misLibros.datoNodo(num);
                    cout << "Titulo: " << libro.getNombre() << "\n";
                    cout << "Autor: " << libro.getAutor() << "\n";
                    cout << "Publicacion: " << libro.getPublicacion() << "\n";
                    cout << "Precio: $" << libro.getPrecio() << "\n";
				} else {
                    cout << "Error, la lista esta vacia.\n";    
				}
                break;
			}
            case(8):{
                cout << "Lista ordenada por Titulo del libro!\n\n";
                misLibros.ordenarTitulo();
                break;
            }
            case(9):{
                cout << "Lista ordenada por Nombre del autor!\n\n";
                misLibros.ordenarAutor();
                break;
            }
            case(10):{
                cout << "Lista ordenada por Precio de menor a mayor!\n\n";
                misLibros.ordenarPrecioAsc();
                break;
            }
            case(11):{
                cout << "Lista ordenada por Precio de mayor a menor!\n\n";
                misLibros.ordenarPrecioDes();
                break;
            }
            case(12): {
                estilo= (estilo)?0:1;
                break;
			}
            case(13): {
                string buscado;
                cout << "Nombre de libro a buscar (coincidencia exacta): ";
                getline(cin, buscado);
                Libro libro= misLibros.buscarLibro_Nombre(buscado);
                if( !libro.getNombre().empty() ){
                    cout << "Titulo: " << libro.getNombre() << "\n";
                    cout << "Autor: " << libro.getAutor() << "\n";
                    cout << "Publicacion: " << libro.getPublicacion() << "\n";
                    cout << "Precio: $" << libro.getPrecio() << "\n";
                }else{
                    cout << "No se encontró el libro \"" << buscado << "\".\n";
                }
                break;
            }
            case(14): {
                string buscado;
                cout << "Nombre del autor del libro (coincidencia exacta): ";
                getline(cin, buscado);
                Libro libro= misLibros.buscarLibro_Autor(buscado);
                if( !libro.getNombre().empty() ){
                    cout << "Titulo: " << libro.getNombre() << "\n";
                    cout << "Autor: " << libro.getAutor() << "\n";
                    cout << "Publicacion: " << libro.getPublicacion() << "\n";
                    cout << "Precio: $" << libro.getPrecio() << "\n";
                }else{
                    cout << "No se encontró ningún libro de \"" << buscado << "\".\n";
                }
                break;
            }
            case(15): {
                string buscado;
                cout << "Nombre de libro a buscar (coincidencia exacta): ";
                getline(cin, buscado);
                ListaDinamica_Libro encontrados= misLibros.buscarLibros_Nombre(buscado);
                if( encontrados.elementos()>0 ){
                    encabezadoTabla();
                    for(int i= 0; i<encontrados.elementos(); i++){
                        Libro libro= encontrados.datoNodo(i);
                        filaTabla(libro);
					}
                    pieTabla();
                }else{
                    cout << "No se encontraron coincidencias\n";
                }
                break;
            }
            case(16): {
                string buscado;
                cout << "Nombre del autor a buscar (coincidencia exacta): ";
                getline(cin, buscado);
                ListaDinamica_Libro encontrados= misLibros.buscarLibros_Autor(buscado);
                if( encontrados.elementos()>0 ){
                    encabezadoTabla();
                    for(int i= 0; i<encontrados.elementos(); i++){
                        Libro libro= encontrados.datoNodo(i);
                        filaTabla(libro);
					}
                    pieTabla();
                }else{
                    cout << "No se encontraron coincidencias\n";
                }
                break;
            }
            case(17): {                
                system("cls");
                cout << "Finalizando programa.\n";
                break;
			}
            default:
                cout << "Opcion no configurada.\n";
		}
        system("pause");
	}while(opc != exit);
    // ---------------------------------------------------------------------------
    return 0;
}